### PR TITLE
[java] Use type inference to resolve complicated cases of CollectionTypeMismatch

### PR DESCRIPTION
**Pull request by oowekyala**
See https://github.com/pmd/pmd/pull/6006#issuecomment-3262152441

This uses type inference to resolve the more complicated cases around generics.

## Related issues

- Followup on #6006

## Ready?

- [x] Missing unit tests for InferenceCtx around ivars with null tvar
- [ ] Missing unit tests for CollectionTypeMismatch for new cases that we can check
- [x] Disable checkstyle MissingBreakInSwitch as our PMD rule NonExhaustiveSwitch is better, it checks for exhaustivity. https://github.com/pmd/build-tools/pull/98
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/internal/infer/IncorporationAction.java [hunk 1]
```diff
@@ -79,24 +89,73 @@ public void apply(InferenceContext ctx) {
...
+                // otherBound <: myBound
+                return checkBound(false, otherBound, myBound, ctx);
+            case EQ:
+                // mybound = otherBound
+                return checkBound(true, myBound, otherBound, ctx);
+            case LOWER:
+                // mybound <: otherBound
+                return checkBound(false, myBound, otherBound, ctx);
+            default:
+                throw AssertionUtil.shouldNotReachHere("exhaustive switch");
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/internal/infer/IncorporationAction.java]: That's the default case demanded by checkstyle, see pmd/build-tools#98

Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/internal/infer/IncorporationAction.java]:
```suggestion
```

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/internal/infer/IncorporationAction.java [hunk 2]
```diff
@@ -79,24 +89,71 @@ public void apply(InferenceContext ctx) {
...
+            case UPPER:
+                // otherBound <: myBound
+                return checkBound(false, otherBound, myBound, ctx);
+            case EQ:
+                // mybound = otherBound
+                return checkBound(true, myBound, otherBound, ctx);
+            case LOWER:
+                // mybound <: otherBound
+                return checkBound(false, myBound, otherBound, ctx);
+            }
```
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/types/internal/infer/IncorporationAction.java:129]:
```suggestion
            }
            throw AssertionUtil.shouldNotReachHere("switch is exhaustive");
```
